File: kevoree-core/org.kevoree.core/src/main/scala/org/kevoree/core/impl/KevoreeCoreBean.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.kevoree.core.impl

import org.kevoree.KevoreeFactory
import org.kevoree.ContainerRoot
import org.kevoree.api.configuration.ConfigurationService
import org.osgi.framework.BundleContext
import org.slf4j.LoggerFactory
import scala.reflect.BeanProperty
import org.kevoree.framework.message._
import scala.actors.Actor
import org.kevoree.api.configuration.ConfigConstants
import org.kevoree.framework._
import org.kevoree.framework.context.KevoreeDeployManager
import deploy.PrimitiveCommandExecutionHelper
import org.kevoree.tools.aether.framework.NodeTypeBootstrapHelper
import org.kevoree.cloner.ModelCloner
import org.kevoree.core.basechecker.RootChecker
import java.util.{UUID, Date}
import org.kevoree.api.service.core.handler.{KevoreeModelUpdateException, UUIDModel, ModelListener, KevoreeModelHandlerService}

class KevoreeCoreBean extends KevoreeModelHandlerService with KevoreeThreadActor {

  @BeanProperty var configService: ConfigurationService = null
  var bundleContext: BundleContext = null;

  def getBundleContext = bundleContext

  def setBundleContext(bc: BundleContext) {
    bundleContext = bc
    KevoreeDeployManager.setBundle(bc.getBundle)
  }


  @BeanProperty var nodeName: String = ""
  @BeanProperty var nodeInstance: AbstractNodeType = null

  var models: scala.collection.mutable.ArrayBuffer[ContainerRoot] = new scala.collection.mutable.ArrayBuffer[ContainerRoot]()
  var model: ContainerRoot = KevoreeFactory.eINSTANCE.createContainerRoot
  var lastDate: Date = new Date(System.currentTimeMillis)
  var currentModelUUID: UUID = UUID.randomUUID();

  def getLastModification = lastDate

  var logger = LoggerFactory.getLogger(this.getClass);
  val modelClone = new ModelCloner

  private def checkBootstrapNode(currentModel: ContainerRoot): Unit = {
    try {
      if (nodeInstance == null) {
        currentModel.getNodes.find(n => n.getName == nodeName) match {
          case Some(foundNode) => {
            val bt = new NodeTypeBootstrapHelper
            bt.bootstrapNodeType(currentModel, nodeName, bundleContext) match {
              case Some(ist: AbstractNodeType) => {
                nodeInstance = ist;
                nodeInstance.startNode()

                //SET CURRENT MODEL 

                model = modelClone.clone(currentModel)
                model.removeAllGroups()
                model.removeAllHubs()
                model.removeAllMBindings()
                model.getNodes.filter(n => n.getName != nodeName).foreach {
                  node =>
                    model.removeNodes(node)
                }
                model.getNodes(0).removeAllComponents()
                model.getNodes(0).removeAllHosts()

              }
              case None => logger.error("TypeDef installation fail !")
            }
          }
          case None => logger.error("Node instance name " + nodeName + " not found in bootstrap model !")
        }
      }
    } catch {
      case _@e => logger.error("Error while bootstraping node instance ", e)
    }
  }

  private def checkUnbootstrapNode(currentModel: ContainerRoot): Option[ContainerRoot] = {
    try {
      if (nodeInstance != null) {
        currentModel.getNodes.find(n => n.getName == nodeName) match {
          case Some(foundNode) => {
            val bt = new NodeTypeBootstrapHelper
            bt.bootstrapNodeType(currentModel, nodeName, bundleContext) match {
              case Some(ist: AbstractNodeType) => {

                val modelTmp = modelClone.clone(currentModel)
                modelTmp.removeAllGroups()
                modelTmp.removeAllHubs()
                modelTmp.removeAllMBindings()
                modelTmp.getNodes.filter(n => n.getName != nodeName).foreach {
                  node =>
                    modelTmp.removeNodes(node)
                }
                modelTmp.getNodes(0).removeAllComponents()
                modelTmp.getNodes(0).removeAllHosts()
                Some(modelTmp)

              }
              case None => logger.error("TypeDef installation fail !"); None
            }
          }
          case None => logger.error("Node instance name " + nodeName + " not found in bootstrap model !"); None
        }
      } else {
        logger.error("node instance is not available on current model !")
        None
      }
    } catch {
      case _@e => logger.error("Error while unbootstraping node instance ", e); None
    }
  }

  private def switchToNewModel(c: ContainerRoot) = {
    //current model is backed-up
    models.append(model)

    // MAGIC NUMBER ;-) , ONLY KEEP 10 PREVIOUS MODEL
    if (models.size > 15) {
      models = models.drop(5)
      logger.debug("Garbage old previous model")
    }

    //Changes the current model by the new model
    model = c
    currentModelUUID = UUID.randomUUID()
    lastDate = new Date(System.currentTimeMillis)

    //TODO ADD LISTENER

    //Fires the update to listeners
    new Actor {
      def act() {
        //NOTIFY LOCAL REASONER
        listenerActor.notifyAllListener()
        //NOTIFY GROUP
        val srs = bundleContext.getServiceReferences(classOf[KevoreeGroup].getName, null)
        if (srs != null) {
          srs.foreach {
            sr =>
              bundleContext.getService(sr).asInstanceOf[KevoreeGroup].triggerModelUpdate()
          }
        }
      }
    }.start()
  }

  override def start: Actor = {
    logger.info("Kevoree Start event : node name = " + configService.getProperty(ConfigConstants.KEVOREE_NODE_NAME))
    setNodeName(configService.getProperty(ConfigConstants.KEVOREE_NODE_NAME));
    super.start()

    //State recovery phase
    /*
    val lastModelssaved = bundleContext.getDataFile("lastModel.xmi");
    if (lastModelssaved.length() != 0) {
      /* Load previous state */
      val model = KevoreeXmiHelper.load(lastModelssaved.getAbsolutePath());
      switchToNewModel(model)
    }  */

    this
  }

  override def stop() {
    logger.warn("Kevoree Core will be stopped !")
    listenerActor.stop()
    super[KevoreeThreadActor].forceStop
    if (nodeInstance != null) {
      try {
        val stopModel = checkUnbootstrapNode(model)
        if (stopModel.isDefined) {
          val adaptationModel = nodeInstance.kompare(model, stopModel.get);
          val deployResult = PrimitiveCommandExecutionHelper.execute(adaptationModel, nodeInstance)
        } else {
          logger.error("Unable to use the stopModel !")
        }

      } catch {
        case _@e => {
          logger.error("Error while unbootstrap ", e)
        }
      }
      try {
        nodeInstance.stopNode()
        nodeInstance == null
      } catch {
        case _@e => {
          logger.error("Error while stopping node instance ", e)
        }
      }
    }
    logger.debug("Kevoree core stopped ")
  }

  val cloner = new ModelCloner
  val modelChecker = new RootChecker

  // treatment of incoming messages
  def internal_process(msg: Any) = msg match {
    //Returns the collection of previous models
    case PreviousModel() => reply(models)
    //Returns a clone of the current system model
    case LastModel() => {
      reply(cloner.clone(model))
    }
    // returns a clone of the current system model with a companion UUID for concurrency management
    case LastUUIDModel() => {
      reply(UUIDModelImpl(currentModelUUID, cloner.clone(model)))
    }
    //Update the system with the model given in parameter, with the UUID for checking
    case UpdateUUIDModel(prevUUIDModel, targetModel) => {
      if (prevUUIDModel.getUUID.compareTo(currentModelUUID) == 0) {
        //TODO CHECK WITH MODEL SHA-1 HASHCODE
        reply(internal_update_model(targetModel))
      } else {
        reply(false)
      }
    }
    //Updates the system to fit to the new model
    case UpdateModel(pnewmodel) => {
      reply(internal_update_model(pnewmodel))
    }
    case _@unknow => logger
      .warn("unknow message  " + unknow.toString + " - sender" + sender.toString + "-" + this.getClass.getName)
  }


  private def internal_update_model(pnewmodel: ContainerRoot): Boolean = {
    if (pnewmodel == null) {
      logger.error("Asking for update with a NULL model !")
      false
    } else {
      try {

        //Model checking
        val checkResult = modelChecker.check(pnewmodel)
        if (!checkResult.isEmpty) {
          logger.error("There is check failure on update model, update refused !")
          import scala.collection.JavaConversions._
          checkResult.foreach {
            cr =>
              logger.error("error=>" + cr.getMessage + ",objects" + cr.getTargetObjects.mkString(","))
          }
          false
        } else {
          //Model check is OK.

          var newmodel = cloner.clone(pnewmodel)
          //CHECK FOR HARA KIRI
          if (HaraKiriHelper.detectNodeHaraKiri(model, newmodel, getNodeName())) {
            logger.warn("HaraKiri detected , flush platform")
            // Creates an empty model, removes the current node (harakiri)
            newmodel = KevoreeFactory.createContainerRoot
            try {
              // Compare the two models and plan the adaptation
              val adaptationModel = nodeInstance.kompare(model, newmodel)
              if (logger.isDebugEnabled) {//Avoid the loop if the debug is not activated
                logger.debug("Adaptation model size " + adaptationModel.getAdaptations.size)
                adaptationModel.getAdaptations.foreach {
                  adaptation =>
                    logger.debug("primitive " + adaptation.getPrimitiveType.getName)
                }
              }
              //Executes the adaptation
              PrimitiveCommandExecutionHelper.execute(adaptationModel, nodeInstance)

              //Cleanup the local runtime
              KevoreeDeployManager.bundleMapping.foreach {
                bm =>
                  try {
                    logger.debug("Try to cleanup " + bm.bundleId + "," + bm.objClassName + "," + bm.name)
                    KevoreeDeployManager.removeMapping(bm)
                    getBundleContext.getBundle(bm.bundleId).uninstall()
                  } catch {
                    case _@e => logger.debug("Error while cleanup platform ", e)
                  }
              }
              logger.debug("Deploy manager cache size after HaraKiri" + KevoreeDeployManager.bundleMapping.size)

              //end of harakiri
              nodeInstance = null

              //place the current model as an empty model (for backup)
              switchToNewModel(newmodel)

              //prepares for deployment of the new system
              newmodel = cloner.clone(pnewmodel)
            } catch {
              case _@e => {
                logger.error("Error while update ", e)
              }
            }
            logger.debug("End HaraKiri")
          }

          //Checks and bootstrap the node
          checkBootstrapNode(newmodel)
          val milli = System.currentTimeMillis
          logger.debug("Begin update model " + milli)
          var deployResult = true
          try {
            // Compare the two models and plan the adaptation
            val adaptationModel = nodeInstance.kompare(model, newmodel);

            //Execution of the adaptation
            deployResult = PrimitiveCommandExecutionHelper.execute(adaptationModel, nodeInstance)
          } catch {
            case _@e => {
              logger.error("Error while update ", e)
              deployResult = false
            }
          }
          if (deployResult) {
            //Merge previous model on new model for platform model
            //KevoreePlatformMerger.merge(newmodel, model)
            switchToNewModel(newmodel)
            logger.info("Deploy result {}", deployResult)
          } else {
            //KEEP FAIL MODEL
            logger.warn("Failed model")
          }
          val milliEnd = System.currentTimeMillis - milli
          logger.debug("End deploy result=" + deployResult + "-" + milliEnd)
          deployResult
        }
      } catch {
        case _@e =>
          logger.error("Error while update", e)
          false
      }
    }
  }

  /**
   * Returns the current model.
   * @Deprecated : Consider using #getLastUUIDModel for concurrency reasons
   */
  @Deprecated
  override def getLastModel: ContainerRoot = {
    (this !? LastModel()).asInstanceOf[ContainerRoot]
  }

  /**
   * Returns the current model with a unique token
   */
  override def getLastUUIDModel: UUIDModel = {
    (this !? LastUUIDModel()).asInstanceOf[UUIDModel]
  }

  /**
   * Asks for an update of the system to the new system described by the model.
   * @Deprecated Consider using #compareAndSwapModel for concurrency reasons
   */
  @Deprecated
  override def updateModel(model: ContainerRoot) {
    this ! UpdateModel(model)
  }

  /**
   * Asks for an update of the system to fit the model given in parameter.<br/>
   * This method is blocking until the update is done.
   * @Deprecated Consider using #atomicCompareAndSwapModel for concurrency reasons.
   */
  override def atomicUpdateModel(model: ContainerRoot) = {
    (this !? UpdateModel(model))
    lastDate
  }

  /**
   * Compares the UUID of the model and the current UUID to verify that no update occurred in between the moment the model had been delivered and the moment the update is asked.<br/>
   * If OK, updates the system and switches to the new model, asynchronously
   */
  override def compareAndSwapModel(previousModel: UUIDModel, targetModel: ContainerRoot) {
    this ! UpdateUUIDModel(previousModel, targetModel)
  }

  /**
   * Compares the UUID of the model and the current UUID to verify that no update occurred in between the moment the model had been delivered and the moment the update is asked.<br/>
   * If OK, updates the system and switches to the new model, synchronously (blocking)
   */
  override def atomicCompareAndSwapModel(previousModel: UUIDModel, targetModel: ContainerRoot): Date = {
    val result = (this !? UpdateUUIDModel(previousModel, targetModel)).asInstanceOf[Boolean]
    if (!result) {
      throw new KevoreeModelUpdateException //SEND AND EXCEPTION - Compare&Swap fail !
    }
    lastDate
  }

  /**
   * Provides the collection of last models (short system history)
   */
  override def getPreviousModel: java.util.List[ContainerRoot] = {
    import scala.collection.JavaConversions._
    (this !? PreviousModel()).asInstanceOf[scala.collection.mutable.ArrayBuffer[ContainerRoot]].toList
  }

  val listenerActor = new KevoreeListeners
  listenerActor.start()

  override def registerModelListener(listener: ModelListener) {
    listenerActor.addListener(listener)
  }

  override def unregisterModelListener(listener: ModelListener) {
    listenerActor.removeListener(listener)
  }


}

File: kevoree-core/org.kevoree.core/src/main/java/org/kevoree/core/impl/KevoreeCoreBean.kt
package org.kevoree.core.impl

import org.kevoree.api.service.core.handler.KevoreeModelHandlerService
import org.kevoree.api.service.core.script.KevScriptEngineFactory
import org.kevoree.core.basechecker.kevoreeVersionChecker.KevoreeNodeVersionChecker
import org.kevoree.ContainerRoot
import org.kevoree.cloner.ModelCloner
import java.util.Date
import org.kevoree.core.basechecker.RootChecker
import java.util.concurrent.ExecutorService
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.ArrayList
import org.kevoree.api.Bootstraper
import java.util.concurrent.Callable
import org.kevoree.api.service.core.handler.UUIDModel
import org.kevoree.api.service.core.handler.ModelUpdateCallback
import org.kevoree.api.service.core.handler.ModelUpdateCallBackReturn
import org.kevoree.api.service.core.handler.ModelHandlerLockCallBack
import java.util.concurrent.ThreadFactory
import org.kevoree.ContainerNode
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import org.kevoree.api.service.core.handler.KevoreeModelUpdateException
import org.kevoree.api.service.core.handler.ModelListener
import org.kevoree.context.ContextRoot
import org.kevoree.framework.HaraKiriHelper
import org.kevoree.framework.KevoreeXmiHelper
import org.kevoree.impl.DefaultKevoreeFactory


class PreCommand(newmodel : ContainerRoot,modelListeners: KevoreeListeners, oldModel : ContainerRoot){
    var alreadyCall = false
    val preRollbackTest: ()->Boolean = {()->
        if (!alreadyCall) {
            modelListeners.preRollback(oldModel, newmodel)
            alreadyCall = true
        }
        true
    }
}

class KevoreeCoreBean(): KevoreeModelHandlerService {

    val modelListeners = KevoreeListeners()
    var _kevsEngineFactory: KevScriptEngineFactory? = null
    private var modelVersionChecker: KevoreeNodeVersionChecker? = null
    var _bootstraper: Bootstraper? = null
    var _nodeName: String = ""
    var nodeInstance: org.kevoree.api.NodeType? = null
    var models: MutableList<ContainerRoot> = ArrayList<ContainerRoot>()
    val kevoreeFactory = org.kevoree.impl.DefaultKevoreeFactory()
    var model: ContainerRoot = kevoreeFactory.createContainerRoot()
    var lastDate: Date = Date(System.currentTimeMillis())
    var currentModelUUID: UUID = UUID.randomUUID()
    var logger = LoggerFactory.getLogger(this.javaClass)!!
    val modelCloner = ModelCloner()
    val modelChecker = RootChecker()
    var selfActorPointer = this
    private var scheduler: ExecutorService? = null
    private var lockWatchDog: ScheduledExecutorService? = null
    private var futurWatchDog: ScheduledFuture<out Any?>? = null
    private var currentLock: LockCallBack? = null

    data class LockCallBack(val uuid: UUID, val callback: ModelHandlerLockCallBack)


    override fun getNodeName(): String {
        return _nodeName
    }

    fun setNodeName(nn: String) {
        _nodeName = nn
        modelVersionChecker = KevoreeNodeVersionChecker(_nodeName)
    }

    override fun getLastModification(): Date {
        return lastDate
    }

    fun setKevsEngineFactory(k: KevScriptEngineFactory) {
        _kevsEngineFactory = k
    }

    fun setBootstraper(b : Bootstraper){
        _bootstraper = b
    }

    private fun switchToNewModel(c: ContainerRoot) {
        var cc: ContainerRoot? = c
        if(!c.isReadOnly()){
            logger.error("It is not safe to store ReadWrite model")
            cc = modelCloner.clone(c, true)
        } else {
            cc = c
        }

        //current model is backed-up
        models.add(model)
        // TODO : MAGIC NUMBER ;-) , ONLY KEEP 10 PREVIOUS MODEL
        if (models.size > 15) {
            models = models.drop(5) as MutableList<ContainerRoot>
            logger.debug("Garbage old previous model")
        }
        //Changes the current model by the new model
        if(cc != null){
            model = cc!!
            currentModelUUID = UUID.randomUUID()
            lastDate = Date(System.currentTimeMillis())
            //Fires the update to listeners
            modelListeners.notifyAllListener()
        }
    }

    override fun getLastModel(): ContainerRoot {
        return scheduler?.submit(LastModelCallable())?.get()!!
    }

    private inner class LastModelCallable(): Callable<ContainerRoot> {
        override fun call(): ContainerRoot {
            return model
        }
    }

    override fun getLastUUIDModel(): UUIDModel {
        return scheduler?.submit(LastUUIDModelCallable())?.get()!!
    }

    private inner class LastUUIDModelCallable(): Callable<UUIDModel> {
        override fun call(): UUIDModel {
            return UUIDModelImpl(currentModelUUID, model)
        }
    }

    inline fun cloneCurrentModel(pmodel: ContainerRoot?) : ContainerRoot {
        return modelCloner.clone(pmodel!!, true)!!
    }

    override fun updateModel(pmodel: ContainerRoot?): Unit {
        scheduler!!.submit(UpdateModelCallable(cloneCurrentModel(pmodel), null))
    }

    inner class UpdateModelCallable(val model: ContainerRoot, val callback: ModelUpdateCallback?): Callable<Boolean> {
        override fun call(): Boolean {
            var res: Boolean = false
             if (currentLock == null) {
                val internalRes = internal_update_model(model)
                callCallBack(callback, internalRes, null)
                res = internalRes
            } else {
                logger.debug("Core Locked , UUID mandatory")
                callCallBack(callback, false, ModelUpdateCallBackReturn.CAS_ERROR)
                res = false
            }
            return res
        }
    }

    fun callCallBack(callback: ModelUpdateCallback?, sucess: Boolean, res: ModelUpdateCallBackReturn?) {
        if (callback != null) {
            object : Thread(){
                override fun run() {
                    if (res == null) {
                        callback?.modelProcessed(if (sucess) {
                            ModelUpdateCallBackReturn.UPDATED
                        } else {
                            ModelUpdateCallBackReturn.DEPLOY_ERROR
                        })
                    } else {
                        callback?.modelProcessed(res)
                    }
                }
            }.start()
        }
    }

    fun start() {
        if (getNodeName() == "") {
            setNodeName("node0")
        }
        modelListeners.start(getNodeName())
        logger.info("Kevoree Start event : node name = " + getNodeName())
        scheduler = java.util.concurrent.Executors.newSingleThreadExecutor(KevoreeCoreThreadFactory(getNodeName()))
    }

    fun stop() {
        logger.warn("Kevoree Core will be stopped !")
        modelListeners.stop()
        scheduler?.shutdownNow()
        scheduler = null
        if (nodeInstance != null) {
            try {
                val stopModel = checkUnbootstrapNode(model)
                if (stopModel != null) {
                    val adaptationModel = nodeInstance!!.kompare(model, stopModel)
                    adaptationModel.setInternalReadOnly()
                    val afterUpdateTest: () -> Boolean = {() -> true }
                    val rootNode = model.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>())!!
                    org.kevoree.framework.deploy.PrimitiveCommandExecutionHelper.execute(rootNode, adaptationModel, nodeInstance!!, afterUpdateTest, afterUpdateTest, afterUpdateTest)
                } else {
                    logger.error("Unable to use the stopModel !")
                }

            } catch (e: Exception) {
                logger.error("Error while unbootstrap ", e)
            }
            try {
                logger.debug("Call instance stop")
                nodeInstance?.stopNode()
                nodeInstance == null
                _bootstraper?.clear()
            } catch(e: Exception) {
                logger.error("Error while stopping node instance ", e)
            }
        }
        logger.debug("Kevoree core stopped ")
    }

    private fun lockTimeout() {
        scheduler?.submit(LockTimeoutCallable())
    }

    inner class LockTimeoutCallable(): Runnable {
        override fun run() {
            if (currentLock != null) {
                currentLock!!.callback.lockTimeout()
                currentLock = null
                lockWatchDog?.shutdownNow()
                lockWatchDog = null
                futurWatchDog = null
            }
        }
    }

    override fun checkModel(tModel: ContainerRoot?): Boolean {
        val checkResult = modelChecker.check(model)
        return if (checkResult!= null && checkResult.isEmpty()!!) {
            modelListeners.preUpdate(model, cloneCurrentModel(tModel))
        } else {
            false
        }
    }

    override fun updateModel(tmodel: ContainerRoot?, callback: ModelUpdateCallback?) {
        scheduler!!.submit(UpdateModelCallable(cloneCurrentModel(tmodel), callback))
    }

    override fun compareAndSwapModel(previousModel: UUIDModel?, targetModel: ContainerRoot?, callback: ModelUpdateCallback?) {
        scheduler?.submit(CompareAndSwapCallable(previousModel!!, cloneCurrentModel(targetModel), callback))
    }

    inner class AcquireLock(val callBack: ModelHandlerLockCallBack, val timeout: Long): Runnable {
        override fun run() {
            if (currentLock != null) {
                callBack.lockRejected()
            } else {
                val lockUUID = UUID.randomUUID()
                currentLock = LockCallBack(lockUUID, callBack)
                lockWatchDog = java.util.concurrent.Executors.newSingleThreadScheduledExecutor()
                futurWatchDog = lockWatchDog?.schedule(WatchDogCallable(), timeout, TimeUnit.MILLISECONDS)
                callBack.lockAcquired(lockUUID)
            }
        }
    }

    inner class WatchDogCallable(): Runnable {
        override fun run() {
            lockTimeout()
        }
    }

    override fun releaseLock(uuid: UUID?) {
        scheduler?.submit(ReleaseLockCallable(uuid!!))
    }

    inner class ReleaseLockCallable(val uuid: UUID): Runnable {
        override fun run() {
            if (currentLock != null) {
                if (currentLock!!.uuid.compareTo(uuid) == 0) {
                    currentLock = null
                    futurWatchDog?.cancel(true)
                    futurWatchDog = null
                    lockWatchDog?.shutdownNow()
                    lockWatchDog = null
                }
            }
        }
    }

    inner class CompareAndSwapCallable(val previousModel: UUIDModel, val targetModel: ContainerRoot, val callback: ModelUpdateCallback?): Callable<Boolean> {
        override fun call(): Boolean {
            val res: Boolean = if (currentLock != null) {
                if (previousModel?.getUUID()?.compareTo(currentLock!!.uuid) == 0) {
                    val internalRes = internal_update_model(targetModel)
                    callCallBack(callback, internalRes, null)
                    internalRes
                } else {
                    logger.debug("Core Locked , bad UUID " + previousModel.getUUID())
                    callCallBack(callback, false, ModelUpdateCallBackReturn.CAS_ERROR)
                    false //LOCK REFUSED !
                }
            } else {
                //COMMON CHECK
                if (previousModel.getUUID()?.compareTo(currentModelUUID) == 0) {
                    //TODO CHECK WITH MODEL SHA-1 HASHCODE
                    val internalRes = internal_update_model(targetModel)
                    callCallBack(callback, internalRes, null)
                    internalRes
                } else {
                    callCallBack(callback, false, ModelUpdateCallBackReturn.CAS_ERROR)
                    false
                }
            }
            //System.gc()
            return res
        }
    }

    override fun compareAndSwapModel(previousModel: UUIDModel?, targetModel: ContainerRoot?): Unit {
        scheduler?.submit(CompareAndSwapCallable(previousModel!!,cloneCurrentModel(targetModel), null))
    }

    override fun atomicUpdateModel(tmodel: ContainerRoot?): Date? {
        scheduler?.submit(UpdateModelCallable(cloneCurrentModel(tmodel), null))?.get()
        return lastDate
    }

    override fun getPreviousModel(): MutableList<ContainerRoot> {
        return scheduler?.submit(GetPreviousModelCallable())?.get() as MutableList<ContainerRoot>
    }

    private inner class GetPreviousModelCallable(): Callable<List<ContainerRoot>> {
        override fun call(): List<ContainerRoot> {
            return models
        }
    }

    override fun atomicCompareAndSwapModel(previousModel: UUIDModel?, targetModel: ContainerRoot?): Date? {
        val result = scheduler?.submit(CompareAndSwapCallable(previousModel!!, cloneCurrentModel(targetModel), null))?.get()!!
        if (!result) {
            throw KevoreeModelUpdateException() //SEND AND EXCEPTION - Compare&Swap fail !
        }
        return lastDate
    }

    override fun registerModelListener(listener: ModelListener?) {
        modelListeners.addListener(listener!!)
    }

    override fun unregisterModelListener(listener: ModelListener?) {
        modelListeners.removeListener(listener!!)
    }

    override fun getContextModel(): ContextRoot {
        return nodeInstance?.getContextModel()!!
    }


    class RELEASE_LOCK(uuid: UUID){
    }

    override fun acquireLock(callBack: ModelHandlerLockCallBack?, timeout: Long?) {
        scheduler?.submit(AcquireLock(callBack!!, timeout!!))
    }

    private fun checkUnbootstrapNode(currentModel: ContainerRoot): ContainerRoot? {
        try {
            if (nodeInstance != null) {
                val foundNode = currentModel.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>())
                if(foundNode != null){
                    val modelTmp = modelCloner.clone(currentModel)!!
                    modelTmp.removeAllGroups()
                    modelTmp.removeAllHubs()
                    modelTmp.removeAllMBindings()
                    for(node in modelTmp.getNodes()){
                        if(node.getName() != getNodeName()){
                            modelTmp.removeNodes(node)
                        }
                    }
                    modelTmp.getNodes().get(0).removeAllComponents()
                    modelTmp.getNodes().get(0).removeAllHosts()
                    return modelTmp
                } else {
                    logger.error("TypeDef installation fail !")
                    return null
                }
            } else {
                logger.error("node instance is not available on current model !")
                return null
            }
        } catch (e: Exception) {
            logger.error("Error while unbootstraping node instance ", e)
            return null
        }
    }

    private fun checkBootstrapNode(currentModel: ContainerRoot): Unit {
        try {
            if (nodeInstance == null) {
                val foundNode = currentModel.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>())
                if(foundNode != null){
                    nodeInstance = _bootstraper?.bootstrapNodeType(currentModel, getNodeName(), this, _kevsEngineFactory!!)
                    if(nodeInstance!=null){
                        nodeInstance?.startNode()
                        //SET CURRENT MODEL
                        model = modelCloner.clone(currentModel)!!
                        model.removeAllGroups()
                        model.removeAllHubs()
                        model.removeAllMBindings()
                        for(node in model.getNodes()){
                           if(node.getName() != getNodeName()){
                               model.removeNodes(node)
                           }
                        }
                        model.getNodes().get(0).removeAllComponents()
                        model.getNodes().get(0).removeAllHosts()
                    } else {
                        logger.error("TypeDef installation fail !")
                    }

                } else {
                    logger.error("Node instance name " + getNodeName() + " not found in bootstrap model !")
                }
            }
        } catch(e: Exception) {
            logger.error("Error while bootstraping node instance ", e)
            logger.debug(_bootstraper?.getKevoreeClassLoaderHandler()?.getKCLDump())
            try {
                nodeInstance?.stopNode()
            } catch(e: Exception) {
            } finally {
                _bootstraper?.clear()
            }
            nodeInstance = null
        }
    }

    fun internal_update_model(proposedNewModel: ContainerRoot): Boolean {

        if (proposedNewModel.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>()) == null) {
            logger.error("Asking for update with a NULL model or node name ("+getNodeName()+") was not found in target model !")
            return false
        }
        try {
            val readOnlyNewModel = proposedNewModel
            val checkResult = modelChecker.check(readOnlyNewModel)!!
            val versionCheckResult = modelVersionChecker?.check(readOnlyNewModel)!!
            if ( checkResult.size > 0 || versionCheckResult.size > 0) {
                logger.error("There is check failure on update model, update refused !")
                for(cr in checkResult) {
                    logger.error("error=>" + cr?.getMessage() + ",objects" + cr?.getTargetObjects().toString())
                }
                for(cr in versionCheckResult) {
                    logger.error("error=>" + cr?.getMessage() + ",objects" + cr?.getTargetObjects().toString())
                }
                return false
            } else {
                //Model check is OK.
                logger.debug("Before listeners PreCheck !")
                val preCheckResult = modelListeners.preUpdate(model, readOnlyNewModel)
                logger.debug("PreCheck result = " + preCheckResult)
                logger.debug("Before listeners InitUpdate !")
                val initUpdateResult = modelListeners.initUpdate(model, readOnlyNewModel)
                logger.debug("InitUpdate result = " + initUpdateResult)
                   if (preCheckResult && initUpdateResult) {
                       var newmodel = readOnlyNewModel
                       //CHECK FOR HARA KIRI
                       var previousHaraKiriModel: ContainerRoot? = null
                       val hkh = HaraKiriHelper()
                       if (hkh.detectNodeHaraKiri(model, readOnlyNewModel, getNodeName())) {
                           logger.warn("HaraKiri detected , flush platform")
                           previousHaraKiriModel = model
                           // Creates an empty model, removes the current node (harakiri)
                           newmodel = checkUnbootstrapNode(model)!!
                           try {
                               // Compare the two models and plan the adaptation
                               val adaptationModel = nodeInstance!!.kompare(model, newmodel)
                               adaptationModel.setInternalReadOnly()
                               if (logger.isDebugEnabled()){
                                   //Avoid the loop if the debug is not activated
                                   logger.debug("Adaptation model size " + adaptationModel.getAdaptations().size())
                                   for(adaptation in adaptationModel.getAdaptations()) {
                                       logger.debug("primitive " + adaptation.getPrimitiveType()?.getName())
                                   }
                               }
                               //Executes the adaptation
                               val afterUpdateTest: ()->Boolean = {()-> true }
                               val rootNode = newmodel.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>())
                               org.kevoree.framework.deploy.PrimitiveCommandExecutionHelper.execute(rootNode!!, adaptationModel, nodeInstance!!, afterUpdateTest, afterUpdateTest, afterUpdateTest)
                               nodeInstance?.stopNode()
                               //end of harakiri
                               nodeInstance = null
                               _bootstraper?.clear() //CLEAR
                               //place the current model as an empty model (for backup)

                               val backupEmptyModel = kevoreeFactory.createContainerRoot()
                               backupEmptyModel.setInternalReadOnly()
                               switchToNewModel(backupEmptyModel)

                               //prepares for deployment of the new system
                               newmodel = readOnlyNewModel
                           } catch(e: Exception) {
                               logger.error("Error while update ", e)
                           }
                           logger.debug("End HaraKiri")
                       }


                       //Checks and bootstrap the node
                       checkBootstrapNode(newmodel)
                       val milli = System.currentTimeMillis()
                       logger.debug("Begin update model " + milli)
                       var deployResult = true
                       try {
                           // Compare the two models and plan the adaptation
                           logger.info("Comparing models and planning adaptation.")

                           val adaptationModel = nodeInstance!!.kompare(model, newmodel)
                           adaptationModel.setInternalReadOnly()
                           //Execution of the adaptation
                           logger.info("Launching adaptation of the system.")
                           val  afterUpdateTest: ()-> Boolean = {()-> modelListeners.afterUpdate(model, newmodel) }

                           val preCmd = PreCommand(newmodel,modelListeners,model)
                           val postRollbackTest: ()->Boolean = {() -> modelListeners.postRollback(model, newmodel);true }
                           val rootNode = newmodel.findByQuery("nodes[" + getNodeName() + "]", javaClass<ContainerNode>())!!
                           deployResult = org.kevoree.framework.deploy.PrimitiveCommandExecutionHelper.execute(rootNode, adaptationModel, nodeInstance!!, afterUpdateTest, preCmd.preRollbackTest, postRollbackTest)
                       } catch(e: Exception) {
                           logger.error("Error while update ", e)
                           deployResult = false
                       }
                       if (deployResult) {
                           switchToNewModel(newmodel)
                           logger.info("Update sucessfully completed.")
                       } else {
                           //KEEP FAIL MODEL, TODO
                           logger.warn("Update failed")
                           //IF HARAKIRI
                           if (previousHaraKiriModel != null) {
                               internal_update_model(previousHaraKiriModel!!)
                               previousHaraKiriModel = null //CLEAR
                           }
                       }
                       val milliEnd = System.currentTimeMillis() - milli
                       logger.debug("End deploy result=" + deployResult + "-" + milliEnd)
                       return deployResult

                   } else {
                       logger.debug("PreCheck or InitUpdate Step was refused, update aborded !")
                       return false
                   }

            }
        } catch (e: Exception) {
            logger.error("Error while update", e)
            return false
        }
    }

}
